### ISSUE TITLE
Inconsistent upscaling causes issues with shaders

### ISSUE THREAD

**Issue by MorsGames** [higan-emu/ares#541]
I was trying out the [quark-shaders](https://github.com/hizzlekizzle/quark-shaders) with Ares v127, but many of the shaders were giving weird results with some of the cores (noticed with SNES, Genesis, and N64).

Well, it turns out that different cores output the image with different upscaling levels, and it's never really consistent. Some seem to only upscale the image on the horizontal axis (I think? I'm not entirely sure), some do it in both, and some don't do any upscaling at all.

I think this was done because some of these consoles have different output modes, but in that case wouldn't it make more sense to change the output resolution based on the mode, or at the very least sending the image to the shader before any upscaling is done? This would make many of the currently broken shaders start working correctly.

It's possible to work around this issue by downscaling the image within the shader manifest file, but the real issue is that the internal upscale amount is different in each core, and as far as I can tell there's no way to set different shaders per core. Not to mention that this workaround would also break all the different resolution modes, like Sonic 2's multiplayer mode and SNES's fake transparency.

**Comment by Screwtapello** [comment]
This behaviour was inherited from higan's high-accuracy GPU emulation. I don't think ares uses high-accuracy emulation *by default*, but it *can*, and perhaps that behaviour is kept for consistency?

> I think this was done because some of these consoles have different output modes, but in that case wouldn't it make more sense to change the output resolution based on the mode, or at the very least sending the image to the shader before any upscaling is done?

The issue is that the output mode can change *during a frame*. For example, you can make a Genesis game where the top half of each frame is 256px wide, while the bottom half is 320px wide. There's various ways you *might* represent that in a way that shaders could make sense of it, but the overall simplest implementation was just to make the overall output width 1280px, since the shader is just going to rescale the result anyway. Genesis pixels drawn in 256px mode become 5 output pixels, while Genesis pixels drawn in 320px mode become 4 output pixels.

**Comment by MorsGames** [comment]
> The issue is that the output mode can change during a frame. 

Oh that makes sense, I hadn't considered that possibility.

In that case it would probably be most ideal to change the resolution to 1280px whenever both modes are used on the same frame instead. I think it would be worth it because as it is right now it's not possible to use the NTSC shader (as well as some of the CRT ones I'd imagine) on all cores without making core specific changes.

For comparison here's how the NTSC shader looks with default resolution settings.
![image](https://user-images.githubusercontent.com/16962745/166166608-44927c6f-fbbb-4bd2-9d86-fdaf03106001.png)
Not much different than how it would look without a shader.

And here's how it should ideally look, after downscaling the horizontal resolution to 25%, and vertical resolution to 50%.
![image](https://user-images.githubusercontent.com/16962745/166166646-c670dd36-b93d-4822-a594-3085c2b84f60.png)

I guess another solution would be to allow setting different shaders per core but there doesn't seem to be any other core dependent setting like that so I can see that going against the nature of the project.

**Comment by Gryzor1363** [comment]
I can confirm this issue is persisting as of latest version 130.1. 
Seems to be especially and quasi-exclusively impacting CRT shaders of the "quark" pack, with the notable exception of CRT-Geom and CRT Simple, following the same pattern as described above  : uneven/incorrect scaling of shader texture with scanlines, resulting in vertical and/or horizontal blocking that grossly distorts the intended effect in fullscreen.

Noteworthy is that every single of the impacted shaders seem to scale correctly in windowed mode, with scaling + aspect correction, adaptive sizing and auto-centering options enabled.

I have attached the Super Mario Bros in its original NES edition as an example, since its dominantly bright and light-green tones are a good benchmark to whatever can be wrong in shading in general. Shader used in that particular sample was CRT-Royale.

![Supr Mario Bros - NES - Windowed](https://user-images.githubusercontent.com/92555103/206045863-fc0d5d80-eb83-46a7-bd17-39cee6a24ece.JPG)
![Supr Mario Bros - NES - Fullscreen](https://user-images.githubusercontent.com/92555103/206045868-543bd406-b4a1-4a77-a351-e451e4e937eb.jpg)

As a side note, on every SNES game I have tested so far (about a dozen of them), some of the shaders such as CRT-Royale render in a nearly unwatchable, anomalous, strongly flickering interlacing even in windowed mode.

Not fully apparent in a still, I thought useful to attach the fullscreen distorsion that behave in a quite similar way to the NES games. Shader used in the fullscreen image : CRT-Lottes-Multipass.

![Supr Metroid - SNES - Fullscreen](https://user-images.githubusercontent.com/92555103/206046794-0fd31b83-5c13-404a-8b9c-88de1913f6bf.jpg)

**Comment by Screwtapello** [comment]
> Noteworthy is that every single of the impacted shaders seem to scale correctly in windowed mode, with scaling + aspect correction, adaptive sizing and auto-centering options enabled.

If a shader ever renders "correctly" for a given core, that's a different issue.

> As a side note, on every SNES game I have tested so far (about a dozen of them), some of the shaders such as CRT-Royale render in a nearly unwatchable, anomalous, strongly flickering interlacing even in windowed mode.

That sounds more like this issue, yes.

**Comment by Gryzor1363** [comment]
> > Noteworthy is that every single of the impacted shaders seem to scale correctly in windowed mode, with scaling + aspect correction, adaptive sizing and auto-centering options enabled.
> 
> If a shader ever renders "correctly" for a given core, that's a different issue.
> 

My bad, actually that remark was meant as a way to provide you with additional information pointing to what situations made the shaders mostly "right" versus all of those where they are almost always broken, and not exclusively stick to what does simply not work in relation to them. Indeed several use cases produced different results and not a "100% broken feature". Context, so to speak. Nonetheless, I stand corrected on the posting rules, thank you.

> > As a side note, on every SNES game I have tested so far (about a dozen of them), some of the shaders such as CRT-Royale render in a nearly unwatchable, anomalous, strongly flickering interlacing even in windowed mode.
> 
> That sounds more like this issue, yes.

Noted. Then I would emphasize on the fact that this would be the one and only case I have been able to reproduce with a CRT-category shader also failing in windowed, as much as it does in fullscreen, to the point that I considered it an exception to the phenomenon I was describing.

And thanks for your quick reply, @Screwtapello

**Comment by MorsGames** [comment]
> I can confirm this issue is persisting as of latest version 130.1. 
> Seems to be especially and quasi-exclusively impacting CRT shaders of the "quark" pack, with the notable exception of CRT-Geom and CRT Simple, following the same pattern as described above  : uneven/incorrect scaling of shader texture with scanlines, resulting in vertical and/or horizontal blocking that grossly distorts the intended effect in fullscreen.

Yeah I feel like this is a slightly different issue than what I described in my original post.

My complaint was that the internal image buffer was upscaled before being passed to the shader, in ways that was not consistent between cores at all, which impacted the usefulness of the whole shaders feature significantly. The issue is given the "bug" label but it's more like a weird quirk of the emulator, and if the emulator is in fullscreen or windowed should be unrelated to it.

The issue you're describing sounds like the final framebuffer (the one _after_ the shader is applied) getting upscaled incorrectly which is an entirely different problem, unless I'm misunderstanding what you're describing.

**Comment by Gryzor1363** [comment]
Thanks for the clarification to both of you. Indeed my post didn't entirely belong to that thread considering we aren't even discussing the same processing phase/pass and only the resulting visual behavior i.e. distorsion/broken effect seem similar in some ways. I simply cobbled "shader" and "issues" together and thought I would contribute, as far as my (very narrow) technical understanding of shaders and their application goes. My final question would be : Where should I move my own observations, assuming of course that they aren't already covered by another ticket ?

Thx

**Comment by Screwtapello** [comment]
I don't think they're covered by another ticket *yet*. Please create one, with the screenshots and description you added to this issue - that's a good example and description of the problem you're having.

**Comment by Gryzor1363** [comment]
No problem, glad I could bring up something useful. Done.

**Comment by aduffey** [comment]
Is there any possibility that the vertical scaling could be decoupled from the horizontal scaling?

Some CRT shaders will look the same regardless of how many times horizontal pixels are duplicated, but if vertical pixels are duplicated they will either look double-scanned (i.e. 2 scanlines per logical vertical pixel) or they will assume the input is interlaced and will output a flickering mess. I don't think there's really any way around that from the point of view of the shader developer. Currently none of the CRT shaders I've tried (v137, librashader) look right on SFC/SNES or MD/Genesis.

If vertical pixel doubling can be avoided, and higher line counts only used for interlaced content, more shaders will be usable.

**Comment by johnnovak** [comment]
Yeah, I've run into the same thing when trying out this emulator for some SNES games and was quite surprised this issue wasn't caught before, to be honest. It's a show-stopper for CRT shaders.

It's pretty essential that the shader sees 256 x 224 if that's the native resolution. If you want to emulate a CRT, virtually all CRT shaders expect one input line in the image for every "emulated scanline".

If you pass 448 lines to the shader, you get double scanning—that completely breaks the CRT effect, suddenly you're kinda emulating a double-scanned VGA monitor 😆 

Integer scaling horizontally is still a problem because it will make the resulting CRT-shaded image too sharp in the horizontal direction, unlike a typical 80s/90s TV.

For what it's worth, changing the texture size on the fly is not a big deal. We do that all the time in DOSBox Staging whenever the program messes around with the CRTC registers or there is an explicit BIOS video mode change. You just generate a new texture between two frames with the new dimensions, it's no big deal. Then all downstream processing must accommodate varying texture sizes, of course, but it's not a hard thing to do.

**Comment by Screwtapello** [comment]
As mentioned previously, games can *and do* change the resolution *within* a frame, after the frame is partially rendered, not just between one frame and the next.

Granted, it's not possible to switch between 240p and 480i within a frame, so one could allocate a 240-line or 480-line texture as needed, and then scanline shaders would be less confused. But that would make the pipeline from the emulation core to the UI more complicated, and still wouldn't completely solve the problem because of the horizontal-resolution issue.

**Comment by johnnovak** [comment]
> As mentioned previously, games can _and do_ change the resolution _within_ a frame, after the frame is partially rendered, not just between one frame and the next.
> 
> Granted, it's not possible to switch between 240p and 480i within a frame, so one could allocate a 240-line or 480-line texture as needed, and then scanline shaders would be less confused. But that would make the pipeline from the emulation core to the UI more complicated, and still wouldn't completely solve the problem because of the horizontal-resolution issue.

Yeah, sure, I appreciate this would complicate things. I guess I'm saying CRT emulation is kinda essential, so all decisions need to take that into account: how would a CRT shader do its job emulating the _real observable output_ a person would see sitting in front of a CRT? Does the shader have all the necessary "info", or would its job become impossible using the input its being fed with?

As I see it, emulating the digital output is only half the picture (pun intended 😆). There was less emphasis on the accurate recreation of the observable image in all emulators a few decades ago, but every home computer and console emulator needs to emulate the _whole signal chain_. Half of it is the digital output coming out of the digital circuitry, the other half is the various digital to analog converters inside the hardware, plus the cables and the CRT display at the end of the chain, outside of the actual console hardware.

So yeah, you could feed the shaders with a fixed texture plus uniforms that might change from frame to frame, but the shader could use those to reconstruct something close to the actual observable experience. Or changing the texture size is an option, but yeah, not on a frame-to-frame basis. Like I said, I'm not a console emulation expert, this is actually quite interesting that certain consoles could do such trickery on a per-frame basis.

**Comment by aduffey** [comment]
> Integer scaling horizontally is still a problem because it will make the resulting CRT-shaded image too sharp in the horizontal direction, unlike a typical 80s/90s TV.

This is not the case for every shader, and it would be nice to at least use some of them.

I would be in favor of at least switching the vertical size of the texture between frames as needed or providing a uniform. In fact, a uniform specifying whether the current field is even or odd would solve some issues with the way shaders currently handle interlacing. That would probably be a whole other discussion, though, including librashader changes.

I looked through the ares source but I didn't know the architecture well enough to really have a grasp on the amount of work it would take. Does the UI (and ruby?) currently expect the size to be constant?

**Comment by chyyran** [comment]
Just FYI any discussion on adding uniforms would have to go through slang-shaders and possibly involve a reference implementation in RetroArch, being the primary runtime under which slang shader presets are used. I would rather not like to fork the ecosystem by implementing librashader only extensions (an exception being `original` scaling, but that originated in RetroArch and got dropped for some reason, I'm hoping it gets picked back up in the future).

That being said librashader already has support for user parameters which can vary every frame via [`filter_chain_set_param`](
https://docs.rs/librashader-capi/latest/librashader_capi/runtime/gl/fn.libra_gl_filter_chain_set_param.html). This mechanism was intended for slang [user parameters](https://github.com/libretro/slang-shaders?tab=readme-ov-file#user-parameter-support) and so is more expensive (every `set_param` call involves a synchronized access and copy; this is such that reading these params when drawing the frame is lock-free) and limited than a proper uniform, but it could be useful for experimentation without needing to preemptively make changes in a librashader fork.
